LibGit2 credential callback testing framework

omus opened the pull request:
Part of #20725. Changes required to get the prompt testing framework functional. Here's a brief breakdown of how this works:

Added a new function `credential_loop` which emulates the how LibGit2 asks for credentials. Having this function allows us to avoid having to talk to a remote server to validate credentials and allows us to restrict the valid credentials. The `credential_loop` function is called inside of a separate process spawned by `challenge_prompt` which supplies programmed responses when encountering a challenge.

At this point the SSH tests from #20725 aren't working. I may include those test in this PR yet.

omus committed:
Create test helper challenge_prompt

Useful for testing Julia functions which prompt for user input. Note
that the fake PTY and new process are necessary to support user input
that uses the `getpass` function.

omus committed:
Test framework for LibGit2 credential callback

omus commented:
Let me know if I should break apart this PR further.

omus committed:
Add LibGit2 SSH credential prompt tests

omus committed:
Add LibGit2 HTTPS credential prompt tests

omus commented:
Appveyor found an [issue](https://ci.appveyor.com/project/JuliaLang/julia/build/1.0.11610) with Windows. Looks like `JL_O_NOCTTY` isn't defined on Windows. I may need to limit the tests to Unix environments or find an alternative.

omus commented:
I'm unsure how we can make a fake TTY for Windows like can for Linux / Mac. I've limited the tests to only run on UNIX based systems. Note that we probably would need to do something similar if we could get a fake TTY on Windows as the prompt challenge text would need to be changed for Windows.

vtjnash commented:
> I'm unsure how we can make a fake TTY for Windows like can for Linux / Mac

We should probably put together a test framework for this. It'll be limited (a process can only have a single TTY open), and not necessarily simple (ConsoleInput has it's own custom API. Also, while input is stream/packet-oriented, output is not. So it can be used for STDIN pretty easily, but it can't be used to test STDOUT).

yuyichao commented on test/libgit2.jl:
```diff
@@ -796,16 +796,158 @@ mktempdir() do dir
+                # credentials. Since we don't control the internals of LibGit2 though they
+                # could also just re-call the credential callback like they do for HTTP.
+                challenges = [
+                    "Passphrase for $valid_p_key:" => "foo\n",
```
Who generate these messages? Are they translatable?

omus replied:
These messages are hardcoded into [base/libgit2/callbacks.jl](https://github.com/JuliaLang/julia/blob/9d960560833ec2cecff99045b1564f2ca08d3e0a/base/libgit2/callbacks.jl#L120)

yuyichao replied:
OK. Should be good enough for now.

omus commented:
Added @Keno to review as he wrote `with_fake_pty` and is probably the most qualified to review the `challenge_prompt` code.

tkelman commented on base/libgit2/callbacks.jl:
```diff
@@ -268,3 +253,52 @@ mirror_cb() = cfunction(mirror_callback, Cint, (Ptr{Ptr{Void}}, Ptr{Void}, Cstri
+function credential_loop(
+    valid_credential::UserPasswordCredential;
+    url="https://github.com/test/package.jl", user="")
+    credential_loop(valid_credential, url, user, 0x000001)
```
looks really strange to have arguments and function body at the same level of indentation

iamed2 replied:
I do:
```julia
function credential_loop(
    valid_credential::UserPasswordCredential;
    url="https://github.com/test/package.jl", 
    user="",
)
    credential_loop(valid_credential, url, user, 0x000001)
```
and the `)` on its own line gives some separation.

tkelman replied:
I usually give line continuations of block openers an extra level of indent to set them off without using up an extra vertical line, but we haven't totally standardized that.

Why is the default `url` set to anything at all here, that default value doesn't strike me as all that useful.

omus replied:
@tkelman do you have another suggestion? I'll update to use @iamed2's suggestion otherwise

omus replied:
Since the `credential_loop` function exists for testing the `credential_callback` function the  `url` is set to be the testing default. I switch `url` from a keyword to a required argument if you think that is more sensible.

tkelman replied:
if this is a testing-only thing, are users or non-testing functions ever going to call it? should it be defined under base/ or test/ ?

omus replied:
Ideally it would be test/ but since the `challenge_prompt` spawns a new Julia process I need to have it someplace I can access. I could put this in the "test/TestHelpers.jl" which would add some additional includes. Personally, I don't think it hurts to have this in base/ and I've found it convenient for debugging.

tkelman replied:
includes or a `-L` flag might also work and be slightly shorter

tkelman commented:
this should also be tested on the buildbots

is PkgDev relying on any of this? It's not great to rename a bunch of types last-minute, they should at least have deprecations

omus commented:
@tkelman [PkgDev does reference](https://github.com/JuliaLang/PkgDev.jl/blob/707a172605667c46896cfee2291568976c0042ca/src/github.jl#L134) the original type `UserPasswordCredentials`. I think I'll separate the type renaming portion of this PR so we can focus on the testing framework.

omus commented:
Revised PR to no longer include renaming of the `UserPasswordCredentials` or `SSHCredentials` types.

tkelman commented:
> Remove libgit2-online incorrect url test

Is it harmful to leave this in, was it causing problems of some kind?

omus commented:
> Is it harmful to leave this in, was it causing problems of some kind?

The test [sometimes was failing](https://ci.appveyor.com/project/JuliaLang/julia/build/1.0.11645/job/pce5unwotoxv6sss#L1865) due to server issues.

tkelman commented:
That's been happening for a long time (#16555), and I think you're misreading it. Looks to me like the issue is in the pkg test, not libgit2-online.

omus commented:
You are correct. I'll make sure to drop that commit

omus committed:
Move credential_loop out of Base

omus commented:
I moved the `credential_loop` function out of Base as it is only needed for testing. Code should be ready for merging.

tkelman commented:
best to run through the buildbots first

omus commented:
> best to run through the buildbots first

No arguments there. I should have said "...ready for merging once builds complete"

tkelman commented:
I just triggered a set of jobs, will see if any issues. There's also staticfloat's jlbuild bot that we could try.

omus commented:
Thanks @tkelman

kshyatt commented:
This is really spectacular, @omus, absolutely gr9 work.

omus commented:
> I just triggered a set of jobs, will see if any issues.

Any news?

tkelman commented on base/libgit2/types.jl:
```diff
@@ -670,32 +670,46 @@ function securezero!(cred::UserPasswordCredentials)
     pass::String
-    pubkey::String
     prvkey::String
+    pubkey::String
```
what's the reasoning for rearranging the fields here? that could be breaking if anything external is constructing this type directly

omus replied:
Not strictly necessary for this PR but makes the field order match the constructor order. Since this type has inner constructors is it even possible to construct this type directly?

tkelman replied:
ah yeah, was there a constructor that took these parameters at all before?

omus replied:
There was no constructor before this PR that let you specify the private and public key.

tkelman commented:
looked okay on buildbots

tkelman commented on base/libgit2/types.jl:
```diff
@@ -670,32 +670,46 @@ function securezero!(cred::UserPasswordCredentials)
     return cred
 end
 
+function Base.:(==)(a::SSHCredentials, b::SSHCredentials)
```
(nit) 8 lines is a little excessive when
```
Base.:(==)(a::SSHCredentials, b::SSHCredentials) = a.user == b.user &&
    a.pass == b.pass && a.prvkey == b.prvkey && a.pubkey == b.pubkey
```
would do

omus replied:
I try to really avoid using the short function syntax when I use need to use multiple lines. Is this ok with you?
```julia
function Base.:(==)(a::SSHCredentials, b::SSHCredentials)
    a.user == b.user && a.pass == b.pass && a.prvkey == b.prvkey && a.pubkey == b.pubkey
end
```

omus committed:
Refactor == for SSHCredentials

omus commented:
Any other comments?

kshyatt commented:
My comment is that this PR is great and thank you for being so patient about all the feedback.

StefanKarpinski commented:
The failure is the usual AppVeyor-can't-connect-to-github issue.

omus commented:
@kshyatt thanks 😄

omus commented:
I'll merge this later today.

tkelman commented on base/libgit2/callbacks.jl:
```diff
@@ -28,16 +28,8 @@ function authenticate_ssh(creds::SSHCredentials, libgit2credptr::Ptr{Ptr{Void}},
-    if errcls != Error.None
-        # Check if we used ssh-agent
-        if creds.usesshagent == "U"
-            creds.usesshagent = "E" # reported ssh-agent error, disables ssh agent use for the future
```
what's the reason for dropping this?

omus replied:
In the [logic that controls the ssh-agent support](https://github.com/JuliaLang/julia/pull/20738/files/b5dfcc7fb095f9ae843777af02771c7723dfb775#diff-43fbc9e635c9a9f0da9707c61e66dea7R32) you could run into an infinite loop if `git_cred_ssh_key_from_agent` would always return 0. To avoid this I dropped `creds.usesshagent == "U"` which made this logic unnecessary.

omus replied:
This change was originally introduced in f5a097c7eb9fe354081e2a8a8b44501980457989. I'll double check that removing this doesn't break this comment:

> Don't skip using the SSH agent if there are two private packages

omus replied:
Removing of this did cause the SSH agent only to be used the first time. I've reintroduced this code and updated the helper code to disable using the SSH agent.

The infinite loop scenario I mentioned only appears when working with `credential_loop` which is only used in the tests. There are checks in place within `credential_loop` to abend if this scenario occurs.

Unfortunately, at this time I cannot create tests that utilize the SSH agent so I cannot write tests for the behaviour I broke. However, I did introduce a comment to increase awareness.

omus committed:
Re-introduce SSH agent error checking

When re-using credentials like we do with `Pkg.update` we need to allow
`creds.usesshagent == "U"` to allow separate authentications to the same
host to both use the SSH agent. This allows the SSH agent to be used
multiple times when we have multiple private packages.

Re-introducing this code did mean we needed to introduce a way of
disabling the use of the SSH agent for testing purposes.

omus commented:
Note that the new tests produce a warning:
```
WARNING: readuntil(IO,AbstractString) will perform poorly with a long string
```
After the feature freeze this will be fixed by #20621

omus committed:
[ci skip] Review change

omus commented:
@tkelman can I get your approval?

tkelman commented:
Running the latest version through the buildbots one last time just to be sure, but yeah looks okay. Would intermediate commits pass or do you plan on squashing?

omus commented:
Intermediate commits should pass tests. My only reason not to squash would be to keep ec49c9eb1b3c6726fbf5c71950250c513b621c8a as a separate commit.

omus commented:
Do the buildbot overlords approve?

oscardssmith commented:
According to [this](https://coveralls.io/repos/25553/builds), codecov decreased 30% for Julia because of this, is it b.c. we need more tests now or b.c. something is broken?

omus commented:
Seeing how this PR mainly just adds test it sounds like something is broken with coverage. Looks like the problem may be fixed by https://github.com/JuliaLang/julia/commit/4e698cb55248b8191c4d68f9c18b656fecf48da4
